ARROW-12208: [C++] Add the ability to run async tasks without using the CPU thread pool

westonpace opened the pull request:
Added a serial executor which can run async code serially without needing to create threads.  In addition, modifies the Scanner::ToTable and FileSystemDataset::Write to execute serially.  Scanner::Scan is still non-serial but I think this is OK.  The I/O thread pool is still being used however.  To truly remove all instances of std::thread we'd need to change the I/O context to use the serial executor but let's see if this makes R happy enough first.

westonpace commented:
@pitrou I initially wanted to replace the IOContext's executor with the serial executor as well but I wasn't sure what the best way to do that would be.  For example, I could create a new IOContext with the same pool & stop token but a different executor and pass that down but I wasn't sure if that would be correct.

I think a simplified version of my question is "does the IOContext come from the calling user (via options) or from the filesystem?"

pitrou commented:
> I think a simplified version of my question is "does the IOContext come from the calling user (via options) or from the filesystem?"

Indeed, I think we should clarify this. I would say: by default, the default IOContext comes from the filesystem but can be overriden by the user. But we may need to modify the IOContext API slightly so that `executor = nullptr` is taken to mean "take the default".

westonpace commented:
Ok, that makes sense.  I think that RTools is happy enough with the I/O threads (they have always existed anyways) so I'll defer that work for the future.

westonpace commented:
JNI check seems unrelated.  Please review.

lidavidm commented on cpp/src/arrow/util/thread_pool.h:
```diff
@@ -262,5 +321,40 @@ class ARROW_EXPORT ThreadPool : public Executor {
+  return SerialExecutor::RunInSerialExecutor<T>(OuterCallback{std::move(get_future)});
+}
+
+/// \brief Potentially runs an async operation serially if use_threads is true
+/// \see RunSerially
+///
+/// If `use_threads` is false then the operation is run normally but this method will
+/// still block the calling thread until the operation has completed.
```
```suggestion
/// \brief Potentially runs an async operation serially if use_threads is false
/// \see RunSerially
///
/// If `use_threads` is true then the operation is run normally but this method will
/// still block the calling thread until the operation has completed.
```
Did you mean this?

westonpace replied:
I used Antoine's comments below which fixes this.

lidavidm commented on cpp/src/arrow/util/thread_pool_test.cc:
```diff
@@ -123,6 +123,125 @@ class AddTester {
+
+TEST(TestSerialExecutor, ContinueAfterExternal) {
+  bool continuation_ran = false;
+  EXPECT_OK_AND_ASSIGN(auto mockIoPool, ThreadPool::Make(1));
```
style nit: mock_io_pool

westonpace replied:
Thanks.  Fixed.

lidavidm reviewed (approved):
I left some tiny nits but this looks good overall.

pitrou commented on cpp/src/arrow/util/thread_pool.cc:
```diff
@@ -44,6 +44,63 @@ struct Task {
+  return Status::OK();
+}
+
+void SerialExecutor::MarkFinished(bool& finished) {
```
`bool* finished`, since it's mutable.

westonpace replied:
Fixed.

pitrou commented on cpp/src/arrow/util/thread_pool.cc:
```diff
@@ -44,6 +44,63 @@ struct Task {
+
+  while (!finished) {
+    while (!state_->task_queue.empty()) {
+      Task& task = state_->task_queue.front();
```
The `task_queue` can be mutated from another thread, and I don't think there are any guarantees about reference stability.

westonpace replied:
True, I thought I was fine since I re-lock before I pop but I suppose the underlying storage could be getting reallocated.

westonpace replied:
Fixed.

pitrou commented on cpp/src/arrow/util/thread_pool.h:
```diff
@@ -189,6 +190,64 @@ class ARROW_EXPORT Executor {
+  template <typename T = ::arrow::detail::Empty>
+  using FinishSignal = internal::FnOnce<void(const Result<T>&)>;
+  template <typename T = ::arrow::detail::Empty>
+  using Scheduler = internal::FnOnce<Status(Executor*, FinishSignal<T>)>;
```
"Scheduler" is confusing. "TopLevelTask" perhaps?

westonpace replied:
Renamed to TopLevelTask

pitrou commented on cpp/src/arrow/util/thread_pool.h:
```diff
@@ -189,6 +190,64 @@ class ARROW_EXPORT Executor {
+/// \brief An executor implementation that runs all tasks on a single thread using an
+/// event loop.
+///
+/// Note: Any sort of nested parallelism will deadlock this executor.  Blocking waits are
```
This is true of a parallel executor as well, no?

westonpace replied:
True, I added the comment to the other thread pool as well.

pitrou commented on cpp/src/arrow/util/thread_pool.h:
```diff
@@ -262,5 +321,40 @@ class ARROW_EXPORT ThreadPool : public Executor {
+/// If `use_threads` is false then the operation is run normally but this method will
+/// still block the calling thread until the operation has completed.
+template <typename T>
+Result<T> RunSynchronously(FnOnce<Future<T>(Executor*)> get_future, bool use_threads) {
```
Why not return a `Future<T>` instead? It seems that would at least remove the `FinishSignal` complication.

westonpace replied:
I'm not sure I follow.  The purpose of this code is to run asynchronous code synchronously in the calling thread.  So the function cannot return until future has been completed.

For example, consider an asynchronous operation that reads a file asynchronously from the filesystem and parses it into a schema.  First...

* The calling thread sets up the read (this happens in the calling thread even if run asynchronously) and then launches the read with io_executor->Submit.
* The I/O thread performs the read.  Meanwhile, the calling thread waits for the read to finish.  This is different than standard asynchronous flow.  In standard asynchronous flow the calling thread would have returned a future at this point.
* The I/O thread finishes the read and adds a follow-up task to the CPU executor.  In this case that is our newly created thread pool.  The calling thread wakes up and processes the follow-up task.  In normal asynchronous flow this follow-up task would have been submitted to the CPU thread pool.
* The calling thread finishes the follow-up task and calls the finish signal.  The calling thread then returns all the way out of the `RunSynchronously` call.  In normal asynchronous flow we would have returned from that call long ago and the CPU thread would just be marking a future complete.

However, there are some other choices we can make here.

* We could require the I/O executor be replaced with our newly created serial executor as well.  This way we don't use any I/O threads OR CPU threads (this is how the previous implementation operated anyways).  In that case we don't need the finish condition because we know we are done when we run out of tasks in the thread pool.  We could also modify the SpawnReal function to return an invalid status if the current calling thread was not the original calling thread (this would safeguard against mistakenly using the I/O executor).
* We could keep track (perhaps in some thread local) of a counter when we schedule tasks on other executors and then decrement this "outgoing task count" when those tasks finish.  In that case we don't need the finish condition because we know we are finished when `outgoing task count == 0 && task_queue.empty()`

pitrou replied:
> So the function cannot return until future has been completed.

I know, but the function can still return the completed future instead of the result. It seems that would remove the callback complications, since `RunLoop` will simply return when everything is finished.

westonpace replied:
Hmm...maybe I'm not clear on what you are saying.

I could change TopLevelTask to return a future and then I could get rid of `FinishSignal`.  That would help simplify the code but wouldn't really change any of the fundamental constraints.

Or are you arguing that `RunSynchronously` should return a finished `Future` instead of a `Result`.  In that case I don't see the advantage since we will still require that the future be finished before we can return and there is not much use with a finished future vs a result.

pitrou replied:
> I could change TopLevelTask to return a future and then I could get rid of FinishSignal. 

That's what I'm saying indeed. It seems one or two layers of callbacks could be removed that way.

westonpace replied:
Ok, yes, this does indeed simplify things.  I've done this.

pitrou commented on cpp/src/arrow/util/thread_pool.h:
```diff
@@ -262,5 +321,40 @@ class ARROW_EXPORT ThreadPool : public Executor {
+  return SerialExecutor::RunInSerialExecutor<T>(OuterCallback{std::move(get_future)});
+}
+
+/// \brief Potentially runs an async operation serially if use_threads is true
```
This isn't very descriptive. Please something such as:
```c++
/// \brief Spawn a future and any dependent tasks on an executor, wait for completion
///
/// If `use_threads` is true, the global CPU executor is used.
/// If `use_threads` is false, a temporary SerialExecutor is used.
/// `get_future` is called (from this thread) with the chosen executor and must
/// return a future that will eventually finish. This function returns once the
/// future has finished.
```

westonpace replied:
Used the suggested comments, thanks.

pitrou commented on cpp/src/arrow/util/thread_pool.h:
```diff
@@ -262,5 +321,40 @@ class ARROW_EXPORT ThreadPool : public Executor {
 // Return the process-global thread pool for CPU-bound tasks.
 ARROW_EXPORT ThreadPool* GetCpuThreadPool();
 
+/// \brief Runs a potentially async operation serially
```
"Run"

westonpace replied:
Fixed.

pitrou commented on cpp/src/arrow/util/thread_pool_test.cc:
```diff
@@ -123,6 +123,125 @@ class AddTester {
+  bool task_ran = false;
+  SerialExecutor::Scheduler<> task = [&](Executor* executor,
+                                         SerialExecutor::FinishSignal<> finish_signal) {
+    EXPECT_TRUE(executor != nullptr);
```
`ASSERT_NE`

westonpace replied:
Fixed.

pitrou commented on cpp/src/arrow/util/thread_pool_test.cc:
```diff
@@ -123,6 +123,125 @@ class AddTester {
   std::vector<int> outs_;
 };
 
+template <typename T = arrow::detail::Empty>
```
Add comments for this?

westonpace replied:
These methods went away addressing the next comment.

pitrou commented on cpp/src/arrow/util/thread_pool_test.cc:
```diff
@@ -123,6 +123,125 @@ class AddTester {
+        std::move(finish_signal)(Status::Invalid("XYZ"));
+        return Status::OK();
+      };
+  ASSERT_RAISES(Invalid, SerialExecutor::RunInSerialExecutor(std::move(scheduler)));
```
Hmm... instead of testing `RunInSerialExecutor` explicitly, can all these tests call `RunSynchronously`, so that you can run the same tests for different `use_threads`?

westonpace replied:
Fixed.

pitrou commented on cpp/src/arrow/dataset/file_csv.cc:
```diff
@@ -166,14 +174,15 @@ class CsvScanTask : public ScanTask {
 
   Result<RecordBatchIterator> Execute() override {
-    ARROW_ASSIGN_OR_RAISE(auto gen, ExecuteAsync());
+    ARROW_ASSIGN_OR_RAISE(auto gen, ExecuteAsync(internal::GetCpuThreadPool()));
```
Shouldn't you lookup `ScanOptions::use_threads`? Or am I missing something?

westonpace replied:
This is a bit awkward.  This method would not be used by either `Scanner::ToTable` or `FileSystemDataset::Write` (they will call `ExecuteAsync`).  The only way this could be called is if the user called `Scan()` directly.

In that case we cannot really use `RunInSerialExecutor` here because it is returning an iterator.

I suppose that means there is no truly serial way to call `Scan()` on CSV.  It wouldn't be parallel exactly, it would just be juggling the work between the calling thread, the I/O executor, and the CPU executor.

pitrou commented on cpp/src/arrow/dataset/scanner.h:
```diff
@@ -32,6 +32,7 @@
 #include "arrow/memory_pool.h"
 #include "arrow/type_fwd.h"
 #include "arrow/util/async_generator.h"
+#include "arrow/util/thread_pool.h"
```
No need to, `internal::Executor` is already declared in `arrow/util/type_fwd.h`.
Also, can you try to remove `async_generator.h`?

westonpace replied:
I tried and failed removing `async_generator.h` before.  The crux of the issue was...

`AsyncGenerator<T>` is a typedef of `std::function<Future<T>()>` and not a proper class in its own right (in fact, `async_generator.h` hides all generator classes behind `MakeXYZGenerator` and so it doesn't actually need to export any types beyond `std::function`.

`std::function<Future<T>()>` requires `<functional>`.

We don't want `<functional>` in a `type_fwd`.

I'm open to suggestions.

pitrou replied:
Perhaps we can use `std::function<Future<T>()>` directly instead of the typedef. @bkietz what do you think?

westonpace replied:
Sorry, I was not clear with my language.  By "no types beyond `std::function`" I meant "no classes or structs".  We would still need to include `async_generator.h` for the functions (e.g. `MakeMappedGenerator`).

So without the typedef you still have the problem that you cannot forward declare the `MakeXyz` functions without including `<functional>` in the forward declaration.

pitrou replied:
But the `MakeXyz` functions are not used in `scanner.h`, are they?

westonpace replied:
Well, you caught me out, they're not all named `MakeXyz` :smile: 

```
      auto scan_fut = CollectAsyncGenerator(std::move(scan_gen))
                          .Then([state, id](const RecordBatchVector& rbs) {
                            state->Emplace(rbs, id);
```

This call to `CollectAsyncGenerator` requires the include.

pitrou replied:
I don't see any call to `CollectAsyncGenerator` in `scanner.h`. Am I missing something?

westonpace replied:
Ah, it is in `scanner.cc`.  My mistake.  Yes, I can move this include to `scanner.cc`.

westonpace replied:
Done.

pitrou commented on cpp/src/arrow/dataset/scanner_internal.h:
```diff
@@ -29,10 +29,12 @@
 #include "arrow/dataset/partition.h"
 #include "arrow/dataset/scanner.h"
 #include "arrow/util/logging.h"
+#include "arrow/util/thread_pool.h"
```
Is this required?

pitrou commented on cpp/src/arrow/util/thread_pool.h:
```diff
@@ -189,6 +190,64 @@ class ARROW_EXPORT Executor {
+  /// always mark a future finished (which can someday be aided by ARROW-12207).
+  template <typename T>
+  static Result<T> RunInSerialExecutor(Scheduler<T> initial_task) {
+    auto serial_executor = std::make_shared<SerialExecutor>();
```
Nit, but this can probably be `SerialExecutor serial_executor;`

westonpace replied:
Fixed.

pitrou reviewed (changes requested):
Nice! I hope we can cut down on the complication by returning `Future<T>`.

pitrou commented on cpp/src/arrow/util/thread_pool.h:
```diff
@@ -189,6 +190,64 @@ class ARROW_EXPORT Executor {
+    return final_result;
+  }
+  void RunLoop(const bool& finished);
+  void MarkFinished(bool& finished);
```
Also, it seems `bool finished` may simply be an internal state member of `SerialExecutor`, e.g.:
```
void RunLoop();
void FinishRunning();
```

(though, is the executor meant to be restartable?)

westonpace replied:
I moved `finished` into the internal state.  The executor is not meant to be restartable at all and should not exist outside of the scope of `RunInSerialExecutor`.  I've moved the constructor for `SerialExecutor` to private to better reflect this.

westonpace commented:
I believe I've addressed all comments.  The only outstanding question is how we want to handle `CsvScanTask::Execute` which I've added some comments to.

pitrou commented on cpp/src/arrow/dataset/file_base.cc:
```diff
@@ -418,14 +418,46 @@ Status WriteNextBatch(WriteState& state, const std::shared_ptr<ScanTask>& scan_t
+      });
+    }
+  }
+  RETURN_NOT_OK(task_group->Finish());
```
@westonpace  Is there a reason why you're not using `FinishAsync` here?

westonpace replied:
Hmm, do you mean add `FinishAsync` to `scan_futs` and then wait for them all at `AllComplete`?

westonpace replied:
Done.

westonpace commented on cpp/src/arrow/util/thread_pool.h:
```diff
@@ -233,15 +233,7 @@ class ARROW_EXPORT SerialExecutor : public Executor {
-          MarkFinished();
-          return st;
-        });
+    final_fut.AddCallback([this](const Result<T>&) { MarkFinished(); });
```
There are some differences between `AddCallback` and `Then` here.  Callbacks are run after the future waiters are notified.  This means that `final_fut.result()` would return before `MarkFinished()` has run.  It "shouldn't" matter since it would probably be bad form to mark the final future complete while there are other tasks lying around on the executor.  However, I wasn't sure it would always be safe.  For example, whomever called `MarkFinished` is going to destruct themselves and it might be nice to ensure that happens before we return from `result()`.

On the other hand, by using `Then` we ensure that `MarkFinished()` has completed executing before `final_fut.result()` returns.

So I'm not certain this change is a problem but I feel it does add risk that we don't need.

westonpace replied:
Actually, I guess we can't return from `RunLoop` until `MarkFinished` has completed so `final_fut.result()` is always going to be called on a completed future either way.  Nevermind.  This should be ok.

pitrou replied:
Yes, that's part of my changes (see comment about locking).

westonpace replied:
Ah, so `wait_for_tasks` could be in the process of being destroyed which is UB given we're still in the middle of a call.

westonpace replied:
Ah, sure enough...
> Notifying while under the lock may nevertheless be necessary when precise scheduling of events is required, e.g. if the waiting thread would exit the program if the condition is satisfied, causing destruction of the notifying thread's condition_variable.

From https://en.cppreference.com/w/cpp/thread/condition_variable/notify_one

westonpace commented:
I believe all new changes have been addressed.

nealrichardson commented:
This seems to fix the CI issues at least, so I'm going to merge. We can follow up next week if necessary.

pitrou commented on cpp/src/arrow/csv/reader.h:
```diff
@@ -26,6 +26,7 @@
 #include "arrow/type.h"
 #include "arrow/type_fwd.h"
 #include "arrow/util/future.h"
+#include "arrow/util/thread_pool.h"
```
`arrow/util/type_fwd.h` was probably sufficient.
